### ISSUE TITLE
MAME2003+ and potential licence violations

### ISSUE THREAD

**Issue by arcadez2003** [libretro/mame2003-plus-libretro#1286]
@mahoneyt944 @grant2258 and anyone else who wants to chime in with an opinion around this subject,

We all know the aims of this project and to achive it we take code from later versions of MAME and backport it here
the closer the MAME version to this build the easier and less problematic it becomes, code prior to 2016 is not up for
grabs so i'll ignore that and focus on code we have used after that date.

The codebase in current MAME bears no resemblance to ours we cannot just take it for use here without in many cases
making drastic modifications and alterations to it, it could be argued since we are more or less rewriting it from scratch
it's ideas not code that we are taking.

As i said earlier im not upto speed on GPL vs BSD etc etc so my thinking could be plain wrong here but since you guys
are here you can always keep me right if im in the wrong here when i suggest the following with regards to the code
submission for Beyond Kung Fu.

The original driver was shared out here https://www.1emulation.com/forums/topic/36615-an-old-rom-dump-and-a-new-driver/

It was then added to MAME shortly after
https://github.com/mamedev/mame/commit/ff210468d2285d6394a48b755fac0203ca6d4c93

Compare the above vs my commits here firstly the driver and then the video
https://github.com/libretro/mame2003-plus-libretro/commit/d8ce7d6db535278b029a2b83165362afb2708bc2#diff-90e73a23f2662862b939a79b148ec2dd6d2705f5c7585cfd8aa3682e6b4ca1d2
https://github.com/libretro/mame2003-plus-libretro/commit/97f097df5404dbd005a1b60ede4dc8b260cc284e

Basically if you do a comparison we're talking chalk and cheese here i've taken no code from MAME the source in question
was shared out freely on a public website taken by me and then extensively rewritten from scratch by myself with some contributions from grant and inserted into this core.

Now part of me thinks since none of that code has been used in it's original form there might be a school of thought that
i've used someones ideas rather than their code directly which means any licence contained in that driver in this case GPL2
is null and void.??

Your opinions around this are more than welcome certainly the above argument is the same one i'd use if there are further
C&D's in the pipeline.

**Comment by mahoneyt944** [comment]
I think if the code was made public prior to mame's addition and you used that early version to create the work for this core. It seems like that would not be a violation.

**Comment by arcadez2003** [comment]
My thinking exactly it's the blitter code which is the problem apparently however once again i'd state it's not in the original form as per that driver as i partially rewrote it and there are many modifications contained within it. again it's open to intepretration whether i've simply taken someones code or someone ideas.

The latter is not licenced as far as im aware :)

**Comment by markwkidd** [comment]
> I think if the code was made public prior to mame's addition and you used that early version to create the work for this core. It seems like that would not be a violation.

I haven't done any of this work, so I don't have any decision-making here, but FWIW I agree with mahoneyt944's quoted statement. I admire y'alls efforts.

**Comment by arcadez2003** [comment]
If they come at us with more requests to remove MAME code post 2016 i'll use the same argument take this commit.....
https://github.com/libretro/mame2003-plus-libretro/commit/b25d57b6faa3d23ddf75fb4c70a8084d79f7afe4

Looks simple enough but there was three stages to this firstly add clipping support as per MAME116 the second fix
comes from MAME138 and required the previous one so i could implement it, both of these are prior to 2016 the final fix however is post 2016.

But again since my code is different in both how i wrote it and how i've added it here once again it could be argued i've 
taken the idea but not the code directly, but if it has to be removed to be on the safe side we still have the earlier fixes
to fall back on this wont be the case everytime but more often than not it will be as a workaround can and will be found.

**Comment by mahoneyt944** [comment]
I don't see a GPL2 header on that file. Might still be safe

**Comment by grant2258** [comment]
I think the best course of action is to deal with it when it comes and if its valid remove it a rework it if you need too @arcadez2003  keep yourself right m8 if you remove it jobs done no drama everyone can move on doesnt need to be painful seems libreto and mame cant get passed the weaponizing the gpl stage. Its actually easier to just do it and stop the constant bs if its a violation everyone sick to death of it. Ask someone at libretro to look at the code and deal with it if its to be added that way you know youve done your part and they can add it back if they choose too.

**Comment by arcadez2003** [comment]
Yeah i know it's just a pity to get caught up in all this and annoying with it as i put in a fair amount of graft with regards to
getting the game up and running, im under the impression i cannot add it back anyhow so that will fail on someone elses
shoulders

**Comment by m4xw** [comment]
> We all know the aims of this project and to achive it we take code from later versions of MAME and backport it here
the closer the MAME version to this build the easier and less problematic it becomes, code prior to 2016 is not up for
grabs so i'll ignore that and focus on code we have used after that date.

No thats wrong. As far I know pre-2016 is the same non commercial license, so its fair game to use it.
Everything 2016+ is a mix of BSD and GPL sources. You only need to be careful with *GPL Only Sources* in 2016+, BSD sources still are fair game no matter what.

> The codebase in current MAME bears no resemblance to ours we cannot just take it for use here without in many cases
making drastic modifications and alterations to it, it could be argued since we are more or less rewriting it from scratch
it's ideas not code that we are taking.

Just changing the macros and throwing codeblocks in the corresponding same funcs is not transformative.

> The original driver was shared out here https://www.1emulation.com/forums/topic/36615-an-old-rom-dump-and-a-new-driver/

That would be fine if the file thats shared there didnt have a big GPL block in it and/or stating a license otherwise 
![image](https://user-images.githubusercontent.com/13141469/152554955-4c95c7b8-678d-4714-907f-05ba63ace3ad.png)

**Comment by m4xw** [comment]
Frankly such things should've never been backported. I can't even blame that on mame. If we (leads) would've noticed it, we would've told you earlier. It's different than the violations they usually meme about (which actually restrict the RA license, not the mame license).
Here it's violating not-our-code tho, so either reach out to the authors for permission under old mame license or spin a clean room implementation.

**Comment by mahoneyt944** [comment]
Being that I haven't looked at the source code could arcadez describe to me how these functions work so I can create my own version?

**Comment by m4xw** [comment]
But yea it will be a significant amount of work to audit the history. I will help on that too, for now stuff can stay until either we verify it or copyright owners push for removal of drivers before we get to check stuff ourselves.
Under different circumstances just widely backporting stuff would've flat out meant the death of the core btw.

**Comment by m4xw** [comment]
> Being that I haven't looked at the source code could arcadez describe to me how these functions work so I can create my own version?

yes

**Comment by m4xw** [comment]
also rom hashes and their maps should be fair game as theres only one solution to the issue

**Comment by mahoneyt944** [comment]
I'll have you review my version before resubmission.

**Comment by arcadez2003** [comment]
Yeah sorry about this i should have paid more attention to licence currantly im checking MAME files with that header
to see if i've taken any code from them for use in this project, hopefully there wont be too many and with some luck
there might even be none we'll see.

**Comment by m4xw** [comment]
> But again since my code is different in both how i wrote it and how i've added it here vs the original code contained
> in MAME again it could be argued i've taken the idea but not the code directly, but if it has to be removed to be on the safe side we still have the earlier fixes to fall back on this wont be the case everytime but more often than not it will be as a workaround can and will be found.

I should clarify tho that technically theres only troubles for one func `kungfum2_state::kungfum2_blitter_w`
The rest falls under triviality for me and that theres only one solution (hashes and mappings)
Also I should note that we (libretro) need to stay conform on that part - we would've brought it up ourselves if we would've been aware its not the usual nonsense violation over our own software.

**Comment by m4xw** [comment]
Scream wolf often enough and nobody listens ;)

**Comment by twinaphex** [comment]
Let's have noone quit over this and let's simply resolve the situation properly.

**Comment by twinaphex** [comment]
@arcadez2003 I understand tensions are high, but we have to be able to show the world we are better than this and not fall to their level. What I will do is force push the last few commits and instead just remove the offending bits without any off hand comments about any one person.

Let's please do this properly and let's not make any more obvious gaffes or offhand comments that will cause conflict. Let's keep it strictly to coding. The personalities don't matter, neither does it matter what they state about us.

**Comment by arcadez2003** [comment]
Ok no problem

**Comment by mahoneyt944** [comment]
@arcadez2003 check discord.

**Comment by twinaphex** [comment]
Let's also further reaffirm that contrary to 'some' people stating otherwise, we will have a full audit of the contributions and where there are issues, they will be rectified.

MAME2003 and MAME2003 Plus are not going anywhere, though. That is a clear line in the sand that will not be crossed, no matter how badly some would want it. Endusers come first.

**Comment by arcadez2003** [comment]
Right i grabbed the latest MAME source and using notepad++ i searched the whole src for the following text 
license:GPL-2.0+ and got 468 hits. i then searched through all the results looking for files that i have taken code
from for inclusion in this core.

Basically there are 5 files in total three of them are fine in my opinion as the code used came long before 
2016 two however will infringe upon GPL2 as part of the code i used was taken from MAME176 which was
released after the 2016 relicence.

@twinaphex @m4xw i can contact one of you privately about this or i can just post details of the code in
question right here for your consideration whichever would be the preferred method.?? obviously i only
know what i've worked on i dont keep track of all the other contributors on here so they'll also have to
check their commits at some point.

**Comment by m4xw** [comment]
Please contact me on the libretro discord, we also need to talk about your .. fiasco.. from yesterday

**Comment by arcadez2003** [comment]
Ok i've decided to give up my contributor status with regards to MAME2003+ i do sincerely apologize for any offense 
i may have caused last night sometimes my temper gets the better of me, however i do not wish for my actions to be
associated with or negatively impact upon the libreto team.

So i think it's best for all concerned that i voluntarily leave this project

Regards.

**Comment by m4xw** [comment]
I wish you the best nontheless

**Comment by DavidHaywood** [comment]
Thank You.  License issues should not be taken lightly, they extend far beyond either project.  If both are to co-exist going forward, these things need to be dealt with properly, and professionally.

In general MAME is moving towards more BSD code, and less GPL code, but that doesn't mean the terms of the GPL can be ignored where it applies, either now or on historical imports made in the period in which these projects have existed.  It is legal obligation by which we are all bound.

This is all part of the process of MAME moving from a highly restrictive license (the one which 0.78 / 0.79 were bound by) to hopefully one with no restrictions at all, the intermediate to that involves some key parts of the code still being GPL however, and means we are not going to refuse submissions that are GPL licensed.

**Comment by Wilstorm** [comment]
@arcadez - I just wanted to say a quick thanks. The pleasure has been mine for the 10's to 100's of hours you plunged into this core. I remember when you shredding solo and adding games like nobody's business to a branch of mame2003 before it became a reformed showcased commodity. I always thought you were the true embodiment of this ideal, in retro gaming (and dumping).

It's not about money, time or prestige. It's not even about preservation, but in its rawest form it's pure nostalgia. It would seem this cores ideologies would align and complement "the other guys" but it's a menacing scene, set against tight fisted mongers and "industry" politics, that view this core as some litany of sins committed on the regular to discompose their harmonies.

Getting along on the back nine I don't get it, at all. I guess people who are unhappy may get some consolation from knowing that others are unhappy too, as they say, misery loves company. Anyway I have great pronounced gratitude to you and your compatriot for breathing new life into an exhausted core to get that feeling one more time, thanks to the Scots (both of ya big yin's...hope I got that right). Thank you...good sirs and I hope you have a great weekend! ;)

**Comment by grant2258** [comment]
@Wilstorm thats the right way. As for the rest well I guess if mame and libretro are happy c'est la vie.

**Comment by mahoneyt944** [comment]
@grant2258 are you on discord? Thought i had an email or something for you but I guess not. Im working on some stuff and wanted to pick your brain if your available

**Comment by grant2258** [comment]
dont use discord and havent sent any mails at all been looking after my cat she passed away today

**Comment by markwkidd** [comment]
Very sorry to hear that Grant. Take care.

**Comment by grant2258** [comment]
cheers mark happend fairly quick thats the things with pets you outlive a few

**Comment by mahoneyt944** [comment]
Sorry to hear.

**Comment by arcadez2003** [comment]
@grant2258 sorry to hear that bigman it's a real shame

**Comment by Wilstorm** [comment]
@grant2258 - Sorry to hear that. I had a girl (cat) that was just short of 20 years before she passed of cancer. I had her through my formative and pubescent years into adulthood. I was in the room when they injected her and it was a pretty "dramatic" experience when they take that last breath. She was around for a lot of my firsts. I was more broken up about that damn cat than half my relatives that passed. :-\
